model/project.go

```diff
@@ -261,23 +262,93 @@ type PluginCommandConf struct {
 	// TimeoutSecs indicates the maximum duration the command is allowed to run for.
 	TimeoutSecs int `yaml:"timeout_secs,omitempty" bson:"timeout_secs"`
 
-	// Params are used to supply configuratiion specific information.
-	Params map[string]interface{} `yaml:"params,omitempty" bson:"params"`
+	// Params are used to supply configuration specific information.
+	// map[string]interface{} is stored as a JSON string.
+	Params map[string]interface{} `yaml:"params" bson:"params"`
+
+	// YAML string of Params to store in database
+	ParamsYAML string `yaml:"params_yaml,omitempty" bson:"params_yaml"`
 
 	// Vars defines variables that can be used within commands.
 	Vars map[string]string `yaml:"vars,omitempty" bson:"vars"`
 
 	Loggers *LoggerConfig `yaml:"loggers,omitempty" bson:"loggers,omitempty"`
 }
 
+func (c *PluginCommandConf) ResolveParams() (map[string]interface{}, error) {
+	out := map[string]interface{}{}
+	if c == nil {
+		return out, nil
+	}
+	if len(c.Params) > 0 {
+		return c.Params, nil
+	}
+	if err := yaml.Unmarshal([]byte(c.ParamsYAML), &out); err != nil {
+		return nil, errors.Wrapf(err, "error unmarshalling params")
+	}
+	c.Params = out
+	return out, nil
+}
+
+func (c *PluginCommandConf) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	temp := struct {
+		Function    string                 `yaml:"func,omitempty" bson:"func"`
+		Type        string                 `yaml:"type,omitempty" bson:"type"`
+		DisplayName string                 `yaml:"display_name,omitempty" bson:"display_name"`
+		Command     string                 `yaml:"command,omitempty" bson:"command"`
+		Variants    []string               `yaml:"variants,omitempty" bson:"variants"`
+		TimeoutSecs int                    `yaml:"timeout_secs,omitempty" bson:"timeout_secs"`
+		Params      map[string]interface{} `yaml:"params,omitempty" bson:"params"`
+		ParamsYAML  string                 `yaml:"params_yaml,omitempty" bson:"params_yaml"`
+		Vars        map[string]string      `yaml:"vars,omitempty" bson:"vars"`
+		Loggers     *LoggerConfig          `yaml:"loggers,omitempty" bson:"loggers,omitempty"`
+	}{}
+
+	if err := unmarshal(&temp); err != nil {
+		return errors.Wrap(err, "error unmarshalling into temp structure")
+	}
+	c.Function = temp.Function
+	c.Type = temp.Type
+	c.DisplayName = temp.DisplayName
+	c.Command = temp.Command
+	c.Variants = temp.Variants
+	c.TimeoutSecs = temp.TimeoutSecs
+	c.Vars = temp.Vars
+	c.Loggers = temp.Loggers
+	c.ParamsYAML = temp.ParamsYAML
+	c.Params = temp.Params
+
+	return c.unmarshalParams()
+}
+
+// we maintain Params for backwards compatibility, but we read from YAML when available, as the
+// given params could be corrupted from the roundtrip
+func (c *PluginCommandConf) unmarshalParams() error {
```

Comment:
should we noop this if c.Params is populated?

Reply:
No--I tried to address in the comments, but basically params could be populated since we are saving it to the database, it would just be corrupted, so we can't rely on it.